homeassistant/components/iskra/sensor.py

```diff
@@ -0,0 +1,343 @@
+                    # If counter is for active energy mark it as energy sensor otherwise dont set device clas.
+                    # This is due to the fact that HA does not support reactive energy sensors.
+                    device_class=(
+                        SensorDeviceClass.ENERGY
+                        if counter.counter_type
+                        in [CounterType.ACTIVE_IMPORT, CounterType.ACTIVE_EXPORT]
+                        else None
+                    ),
+                    # Use HA energy unit for active energy counters otherwise use units from API.
+                    native_unit_of_measurement=(
+                        UnitOfEnergy.WATT_HOUR
+                        if counter.counter_type
+                        in [CounterType.ACTIVE_IMPORT, CounterType.ACTIVE_EXPORT]
+                        else counter.units
+                    ),
+                    value_func=non_resettable_value_func(device, index),
+                )
+                entities.append(
+                    IskraSensor(
+                        coordinator, root_device, entry, sensor_entity_description
+                    )
+                )
+
+            # Add resettable counters.
+            for index, counter in enumerate(device.counters.resettable):
+
+                def resettable_value_func(device, counter_index):
+                    return lambda device: round(
+                        device.counters.resettable[counter_index].value, 2
+                    )
+
+                sensor_entity_description = IskraSensorEntityDescription(
+                    key=f"resettable_counter{index+1}",
+                    translation_key=f"resettable_counter{index+1}",
+                    state_class=SensorStateClass.TOTAL_INCREASING,
+                    # If counter is for active energy mark it as energy sensor otherwise dont set device clas.
+                    # This is due to the fact that HA does not support reactive energy sensors.
+                    device_class=(
+                        SensorDeviceClass.ENERGY
+                        if counter.counter_type
+                        in [CounterType.ACTIVE_IMPORT, CounterType.ACTIVE_EXPORT]
+                        else None
+                    ),
+                    # Use HA energy unit for active energy counters otherwise use units from API.
+                    native_unit_of_measurement=(
+                        UnitOfEnergy.WATT_HOUR
+                        if counter.counter_type
+                        in [CounterType.ACTIVE_IMPORT, CounterType.ACTIVE_EXPORT]
+                        else counter.units
+                    ),
+                    value_func=resettable_value_func(device, index),
+                )
+
+                entities.append(
+                    IskraSensor(
+                        coordinator, root_device, entry, sensor_entity_description
+                    )
+                )
```

Comment:
This looks very complex, can we maybe move this to a follow up so we can focus on the normal devices first?

Reply:
I agree, but this is a crucial aspect of our measuring devices. It is a bit complicated, so let me explain it further.

All our devices, including energy meters and power quality analyzers, have energy counters that measure electrical energy. These devices contain a few non-resettable energy counters, which users cannot reset, and a few resettable ones. This is why there are two `for` loops in the code.

### Resettable Counters
- **Customizable by User**: Users can configure each resettable counter to measure:
  - Imported or exported energy
  - Active, reactive, or apparent energy
  - Total energy or energy for a specific phase of the meter

- **Example Use Case**: This customization can be particularly handy. For instance, a user can set a load such as a central heat pump on one phase and a sanitary heat pump on another phase. They can then see the consumption of each load with one energy meter, as well as the total consumption for the whole heating system.

### Non-resettable Counters
- **Customizable by Buyer**: Larger companies or distributors that purchase from us can customize these counters. However, users cannot reconfigure them during the product's lifetime.

The code checks if the counter type is active energy and, if so, sets the device class to Home Assistant's `SensorDeviceClass.Energy` and the unit of energy to `Watt_HOUR`. Otherwise, it uses the API's units. This ensures compatibility with the Energy tab in Home Assistant.

Reply:
At this moment it's still a blur for me as the code isn't that readable and I can't completely link to whats going on. I don't think we can make 2024.8 (for which the beta cut is wednesday), so let's aim for 2024.9. so in theory we have enough time and I think it would speed up the process to get the way we fetch data and how we create entities straight in this PR and add the extra logic in followups (and then we can do it in understandable bit size pieces). This way we can go over each subject and decide on the best way to do it

Reply:
@joostlek so how do i proceed.  should i remove this part of code?

Reply:
For now I think that's best yes

Reply:
ok removed this part